core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
+            if (isStageFlushing()) {
                 stateMachine.transitionToFlushing();
             }
-            if (finishedTasks.containsAll(allTasks)) {
+            if (isStageFinished()) {
                 stateMachine.transitionToFinished();
             }
         }
     }
 
+    private synchronized boolean isStageFlushing()
+    {
+        // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
+        return !flushingTasks.isEmpty()
+                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+    }
+
+    private synchronized boolean isStageFinished()
+    {
+        return finishedTasks.containsAll(allTasks);
```

Comment:
Yes, you right. `containsAll` is just more readable and it was used before.

Reply:
containsAll is O(N) and we know that this cannot succeed if the source set size is bigger than the target set. With the current implementation, we are looping through allTask twice. Avoiding this could reduce the time needed in the synchronized section.